Checkout: Don't use localStorage to save product data for the thank you page

drewblaisdell opened the pull request:
Fixes #2235.

This PR updates the checkout process, and:
- Changes the route for the checkout thank you page from `/checkout/thank-you` to `/checkout/:site_slug/:receipt_id/thank-you`.
- Adds actions/a reducer/selector to fetch a receipt from the server and store it in the global state tree.
- Dispatches an action to store a receipt when a transaction is completed successfully.
- Fetches a receipt on the thank you page.
- Updates the thank you page to render using purchases from the receipt, instead of items from the last cart.
- Removes `lastTransaction` from local storage, and any logic that used to use that property.

#### Testing instructions - Credit card

1. Run `git checkout fix/thank-you-page` and start your server
2. Open the [`Plans` page](http://calypso.localhost:3000/plans/:site)
3. Purchase a plan with a credit credit card
4. Assert that you are redirect to the thank you page with the correct information on it
5. Navigate to another page (like reader)
6. Assert that clicking back reloads the thank you page

#### Testing instructions - PayPal

1. Run `git checkout fix/thank-you-page` and start your server
2. Open the [`Plans` page](http://calypso.localhost:3000/plans/:site)
3. Purchase a plan with paypal
4. Go through the paypal flow
4. Assert that you are redirect to the thank you page with the correct information on it
5. Navigate to another page (like reader)
6. Assert that clicking back reloads the thank you page

#### Testing instructions - generic thank you page
1. Visit `/checkout/:site/thank-you` for any site.
2. Assert that you see a [generic thank you page](https://cloudup.com/cpStlBbu7LJ).

#### Additional notes

Also worth testing:
- different plans
- premium themes
- domains

#### Reviews

- [x] Code
- [x] Product

drewblaisdell commented:
I broke this into two PRs (the other, #3294, is in review) to make it easier to review. This will be ready for review once:

- I add a placeholder state to the thank you page.
- #3294 is merged.
- This is rebased.

drewblaisdell commented:
This has been rebased, and a loading state for `CheckoutThankYou` has been added.

This should work on the production API for credit card purchases now. Once the API is updated to redirect users that purchase with PayPal back to `/checkout/:site_slug/:receipt_id/thank-you`, this will work for all users.

stephanethomas commented on client/lib/products-values/index.js:
```diff
@@ -281,6 +304,8 @@ module.exports = {
 	isGoogleApps,
+	isJetpackBusiness,
+	isJetpackPremium,
 	isJetpackPlan,
```
The order is almost right :).

stephanethomas commented on client/my-sites/upgrades/index.js:
```diff
@@ -276,8 +276,8 @@ module.exports = function() {
 		page(
-			'/checkout/thank-you',
-			upgradesController.redirectIfThemePurchased,
+			'/checkout/:domain/:receipt_id/thank-you',
```
We should use `:site` instead of `:domain` in the route.

stephanethomas replied:
We should also use `:receiptId` instead of `:receipt_id`.

stephanethomas commented:
I tried to purchase a domain mapping but the following error was raised when displaying the `Thank You` page:

![screenshot](https://cloud.githubusercontent.com/assets/594356/13082276/4f179ac2-d4d0-11e5-918a-fda45cc3a3d3.png)

That's because the [`isDomainMapping()` function](https://github.com/Automattic/wp-calypso/blob/fix/thank-you-page/client/my-sites/upgrades/checkout/thank-you.jsx#L263) expects a `product_slug` property whereas we feed it with a `productSlug`:

```
freeTrial: false
isDomainRegistration: false
meta: "example.com"
productId: 5
productNameShort: null
productSlug: "domain_map"
```

stephanethomas commented:
My understanding was that we shouldn't have to update the API for PayPal since the endpoint already accepts a [success url](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/upgrades/checkout/paypal-payment-box.jsx#L78) (i.e. a return url). However I tried to purchase with PayPal but the following error was raised upon submitting the form:

![screenshot](https://cloud.githubusercontent.com/assets/594356/13082728/331fd742-d4d2-11e5-8a57-8cc270c0ede1.png)

That's because we call [`getCheckoutCompleteRedirectPath()`](https://github.com/Automattic/wp-calypso/blob/fix/thank-you-page/client/my-sites/upgrades/checkout/checkout.jsx#L136) too early during the checkout process, i.e. at a time where no receipt has been fetched yet.

stephanethomas commented:
As mentioned internally, after much head-scratching this afternoon, it seems there is no other solution than hacking the API code to return the new success url.

drewblaisdell commented:
Thanks for looking, @stephanethomas.

> That's because the [`isDomainMapping()` function](https://github.com/Automattic/wp-calypso/blob/fix/thank-you-page/client/my-sites/upgrades/checkout/thank-you.jsx#L263) expects a `product_slug` property whereas we feed it with a `productSlug`:

`isDomainMapping` accepts an object with either camel/snake cased properties:

```js
isDomainMapping( { productSlug: 'domain_map' } ); // => true
```

The problem was that I never imported `find` (:sweat_smile:) and the error message wasn't helpful because `find` is also a [method on `window`](https://developer.mozilla.org/en-US/docs/Web/API/Window/find).

stephanethomas commented on client/my-sites/upgrades/checkout/checkout.jsx:
```diff
@@ -118,8 +120,22 @@ const Checkout = React.createClass( {
 	getCheckoutCompleteRedirectPath: function() {
-		var renewalItem;
+		var renewalItem,
+			receiptId = ':receipt_id';
```
Sorry to be nitpicky but this should be `:receiptId`.

stephanethomas commented on client/my-sites/upgrades/index.js:
```diff
@@ -276,8 +276,8 @@ module.exports = function() {
 		page(
-			'/checkout/thank-you',
-			upgradesController.redirectIfThemePurchased,
+			'/checkout/:site/:receipt_id/thank-you',
```
This was already mentioned before but we should also use `:receiptId` instead of `:receipt_id`.

scruffian commented:
I saw this when testing:

<img width="588" alt="screen shot 2016-02-19 at 10 59 07" src="https://cloud.githubusercontent.com/assets/275961/13173854/cedfa7d0-d6f7-11e5-9664-8709aca98f19.png">

fabianapsimoes commented:
Product :+1: 

I've tested the following products: plans, domain registration, domain mapping, Google Apps and a Premium theme. Both credit card and PayPal.

Nice job, Drew!

stephanethomas commented on client/my-sites/upgrades/checkout/checkout.jsx:
```diff
@@ -133,8 +149,17 @@ const Checkout = React.createClass( {
+				receiptId: receiptId,
+				purchases: this.getPurchasesFromReceipt()
+			} );
+		}
```
I'm not a big fan of accessing the transaction data directly like this. It would be great to hide that logic by using the receipts reducer. To be more specific, what I suggest is to update the `store-transactions` module to update the [receipts state tree](https://github.com/Automattic/wp-calypso/tree/master/client/state/receipts) when [receiving a response](https://github.com/Automattic/wp-calypso/blob/master/client/lib/store-transactions/index.js#L172) from the API. What do you think? That way we would also reduce the number of lines of code by reusing the [receipt assembler](https://github.com/Automattic/wp-calypso/blob/master/client/state/receipts/assembler.js).

drewblaisdell replied:
I opened an issue for this: https://github.com/Automattic/wp-calypso/issues/3422

stephanethomas commented:
I've raised a point about simplify our code but this is something that can be addressed in a different pull request - that way we can deploy this one today.

drewblaisdell commented:
Thanks for the reviews/commits, @fabianapsimoes @stephanethomas @scruffian.

I just realized that this PR also needed a commit that renders a generic thank you page at `/checkout/:site/thank-you`: https://cloudup.com/cpStlBbu7LJ This is necessary for the case where the API redirects the user to that route when we fail to generate a receipt ID (which @peterbutler claims happens a very small percentage of the time).

This is easy to test and a component already existed for it, so I added it in with [this](https://github.com/Automattic/wp-calypso/commit/5b36788c7419c8b8707c6fcc676ad93148d9ceb7) commit.

**Testing**
- Visit `/checkout/:site/thank-you` for any site.
- Assert that you see a [generic thank you page](https://cloudup.com/cpStlBbu7LJ).

I've updated the testing instructions above and unchecked code/product. We should only need to review this commit.

scruffian commented:
Tested and reviewed again :ship:
